File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
from collections import deque
import sys
input = sys.stdin.readline

def bfs():
    q = deque()
    q.append((0, 0, 0, 0))  # x좌표, y좌표, 거리, 대각선으로 움직인 수
    visited[0][0][0] = 1

    while q:
        s = q.popleft()
        x = s[0]
        y = s[1]
        dist = s[2]
        k = s[3]

        if x == H - 1 and y == W - 1:
            return dist

        if k < K:
            for i in range(8):
                nx = x + dxK[i]
                ny = y + dyK[i]
                if 0 <= nx < H and 0 <= ny < W:
                    if not visited[k + 1][nx][ny] and not board[nx][ny]:
                        q.append([nx, ny, dist + 1, k + 1])
                        visited[k + 1][nx][ny] = 1

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < H and 0 <= ny < W:
                if not visited[k][nx][ny] and not board[nx][ny]:
                    q.append([nx, ny, dist + 1, k])
                    visited[k][nx][ny] = 1
    return -1


K = int(input())
W, H = map(int, input().split())

board = [list(map(int, input().split())) for _ in range(H)]

visited = [[[0]*W for _ in range(H)] for _ in range(31)]


dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

dxK = [-2, -1, -2, -1, 1, 2, 1, 2]
dyK = [-1, -2, 1, 2, -2, -1, 2, 1]

print(bfs())

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    static int[] dxK = {-2, -1, -2, -1, 1, 2, 1, 2};
    static int[] dyK = {-1, -2, 1, 2, -2, -1, 2, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int K = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int W = Integer.parseInt(st.nextToken());
        int H = Integer.parseInt(st.nextToken());

        int[][] board = new int[H][W];

        for (int i = 0; i < H; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < W; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][][] visited = new int[31][H][W];
        System.out.println(bfs(K, W, H, board, visited));
    }

    static int bfs(int K, int W, int H, int[][] board, int[][][] visited) {
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, 0, 0, 0});  // x좌표, y좌표, 거리, 대각선으로 움직인 수
        visited[0][0][0] = 1;

        while (!q.isEmpty()) {
            int[] s = q.poll();
            int x = s[0];
            int y = s[1];
            int dist = s[2];
            int k = s[3];

            if (x == H - 1 && y == W - 1) {
                return dist;
            }

            if (k < K) {
                for (int i = 0; i < 8; i++) {
                    int nx = x + dxK[i];
                    int ny = y + dyK[i];
                    if (0 <= nx && nx < H && 0 <= ny && ny < W) {
                        if (visited[k + 1][nx][ny] == 0 && board[nx][ny] == 0) {
                            q.offer(new int[]{nx, ny, dist + 1, k + 1});
                            visited[k + 1][nx][ny] = 1;
                        }
                    }
                }
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (0 <= nx && nx < H && 0 <= ny && ny < W) {
                    if (visited[k][nx][ny] == 0 && board[nx][ny] == 0) {
                        q.offer(new int[]{nx, ny, dist + 1, k});
                        visited[k][nx][ny] = 1;
                    }
                }
            }
        }

        return -1;
    }
}